Framework/CurveFitting/src/Algorithms/DoublePulseFit.cpp

```diff
@@ -0,0 +1,405 @@
+                            const std::string &suffix) {
+  fit.setProperty("InputWorkspace" + suffix, workspace);
+
+  int workspaceIndex = qensFit.getProperty("WorkspaceIndex" + suffix);
+  fit.setProperty("WorkspaceIndex" + suffix, workspaceIndex);
+
+  double startX = qensFit.getProperty("StartX" + suffix);
+  double endX = qensFit.getProperty("EndX" + suffix);
+  fit.setProperty("StartX" + suffix, startX);
+  fit.setProperty("EndX" + suffix, endX);
+
+  std::vector<double> exclude = qensFit.getProperty("Exclude" + suffix);
+  fit.setProperty("Exclude" + suffix, exclude);
+}
+
+void setMultiDataProperties(
+    const Mantid::API::IAlgorithm &qensFit, Mantid::API::IAlgorithm &fit,
+    const std::vector<Mantid::API::MatrixWorkspace_sptr> &workspaces) {
+  setMultiDataProperties(qensFit, fit, workspaces[0], "");
+
+  for (auto i = 1u; i < workspaces.size(); ++i)
+    setMultiDataProperties(qensFit, fit, workspaces[i],
+                           "_" + std::to_string(i));
+}
+
+Mantid::API::IFunction_sptr getDoublePulseFunction(
+    std::shared_ptr<const API::ParamFunction> const &function, double offset,
+    double firstPulseWeight, double secondPulseWeight) {
+  auto clonedFunction = function->clone();
+
+  auto convolution =
+      std::make_shared<Mantid::CurveFitting::Functions::Convolution>();
+  auto delta1 =
+      std::make_shared<Mantid::CurveFitting::Functions::DeltaFunction>();
+  auto delta2 =
+      std::make_shared<Mantid::CurveFitting::Functions::DeltaFunction>();
+  auto deltaComposite = std::make_shared<Mantid::API::CompositeFunction>();
+
+  convolution->setAttributeValue("FixResolution", false);
+  delta1->setParameter("Centre", -offset);
+  delta1->setParameter("Height", firstPulseWeight);
+  delta1->fixAll();
+  delta2->setParameter("Centre", 0.0);
+  delta2->setParameter("Height", secondPulseWeight);
+  delta2->fixAll();
+  deltaComposite->addFunction(delta1);
+  deltaComposite->addFunction(delta2);
+
+  convolution->addFunction(clonedFunction);
+  convolution->addFunction(deltaComposite);
+
+  return convolution;
+}
+
+Mantid::API::IFunction_sptr getDoublePulseFunction(
+    std::shared_ptr<const API::MultiDomainFunction> const &initialFunction,
+    double offset, double firstPulseWeight, double secondPulseWeight) {
+  auto doublePulseFunc = std::make_shared<API::MultiDomainFunction>();
+  for (size_t domain = 0; domain < initialFunction->getNumberDomains();
+       domain++) {
+    auto innerFunction = std::dynamic_pointer_cast<API::ParamFunction>(
+        initialFunction->getFunction(domain));
+    if (innerFunction) {
+      auto twoPulseInnerFunction = getDoublePulseFunction(
+          innerFunction, offset, firstPulseWeight, secondPulseWeight);
+      doublePulseFunc->addFunction(twoPulseInnerFunction);
+      doublePulseFunc->setDomainIndex(domain, domain);
+    } else {
+      throw std::runtime_error("Convolution does not support composite "
```

Comment:
why can we not support composite functions?

Reply:
Currently the first item in a convolution has to have Function1D as a base class. This is because the convolution operator cannot cope with 2D functions. It checks this by checking that the first function of the convolution inherits from Function1D. Despite being 1D functions composite functions which contain only 1D functions fail this check so cannot currently be used as the first element in a convolution. So currently since we have a composite functions of delta's we need the other function to be a single function to work with convolution.

I believe this is a flaw in convolutions design because it fails to acknowledge that composite functions can be 1D. I am planning to try and fix this within the fitting framework but felt it was out of scope for this PR. Sorry meant to explain this in the description but slipped my minid.

Reply:
can you link the issue where you plan to fix it and add a note on it to update this and the QENS code?

Reply:
This shouldn't affect any of the QENS code have created the issue #28880 to resolve it for muons.